localstack-core/localstack/services/events/archive.py

```diff
@@ -0,0 +1,167 @@
+        event_source_arn: Arn,
+        description: ArchiveDescription,
+        event_pattern: EventPattern,
+        retention_days: RetentionDays,
+    ):
+        self.archive = Archive(
+            archive_name,
+            region,
+            account_id,
+            event_source_arn,
+            description,
+            event_pattern,
+            retention_days,
+        )
+        self.set_state(ArchiveState.CREATING)
+        self.set_creation_time()
+        self.client: BaseClient = self._initialize_client()
+        self.event_bus_name = extract_event_bus_name(event_source_arn)
+
+        self.rule_name = self._create_archive_rule()
+        self.target_id = self._create_archive_target()
+        self.set_state(ArchiveState.ENABLED)
+
+    def __getattr__(self, name):
+        return getattr(self.archive, name)
+
+    @property
+    def archive_name(self) -> ArchiveName:
+        return self.archive.name
+
+    @property
+    def archive_arn(self) -> Arn:
+        return self.archive.arn
+
+    def set_state(self, state: ArchiveState) -> None:
+        self.archive.state = state
+
+    def set_creation_time(self) -> None:
+        self.archive.creation_time = datetime.now(timezone.utc)
+
+    def update(
+        self,
+        description: ArchiveDescription,
+        event_pattern: EventPattern,
+        retention_days: RetentionDays,
+    ) -> None:
+        self.set_state(ArchiveState.UPDATING)
+        if description is not None:
+            self.archive.description = description
+        if event_pattern is not None:
+            self.archive.event_pattern = event_pattern
+        if retention_days is not None:
+            self.archive.retention_days = retention_days
+        self.set_state(ArchiveState.ENABLED)
+
+    def delete(self) -> None:
+        # TODO handle active replays
+        self.set_state(ArchiveState.DISABLED)
+        try:
+            self.client.remove_targets(
+                Rule=self.rule_name, EventBusName=self.event_bus_name, Ids=[self.target_id]
+            )
+        except Exception as e:
+            LOG.debug(f"Target {self.target_id} could not be removed, {e}")
+        try:
+            self.client.delete_rule(Name=self.rule_name, EventBusName=self.event_bus_name)
+        except Exception as e:
+            LOG.debug(f"Rule {self.rule_name} could not be deleted, {e}")
+
+    def put_events(self, events: FormattedEventList) -> None:
+        for event in events:
+            self.archive.events[event["id"]] = event
+
+    def get_events(self, start_time: Timestamp, end_time: Timestamp) -> FormattedEventList:
+        events_to_replay = self._filter_events_start_end_time(start_time, end_time)
+        return events_to_replay
+
+    def _initialize_client(self) -> BaseClient:
+        client_factory = connect_to(aws_access_key_id=self.account_id, region_name=self.region)
+        client = client_factory.get_client("events")
+
+        service_principal = ServicePrincipal.events
```

Comment:
because its an internal call setting the service principal to events

Reply:
Probably possible via CloudTrail, but it should be fine unless there's something else to consider for IAM policy enforcement/streams @dfangl